Objects/frameobject.c

```diff
@@ -41,12 +41,19 @@ int
 PyFrame_GetLineNumber(PyFrameObject *f)
 {
     assert(f != NULL);
-    if (f->f_lineno != 0) {
-        return f->f_lineno;
+    if (f->f_lineno == -1) {
+        // We should calculate it once. If we can't get the line number,
+        // set f->f_lineno to 0.
+        f->f_lineno = PyUnstable_InterpreterFrame_GetLine(f->f_frame);
+        if (f->f_lineno < 0) {
+            f->f_lineno = 0;
+        }
     }
-    else {
-        return PyUnstable_InterpreterFrame_GetLine(f->f_frame);
+
+    if (f->f_lineno > 0) {
```

Comment:
If `f->f_lineno == 0` we end up recalculating the line number, so if `PyUnstable_InterpreterFrame_GetLine()` returns -1 it will get called twice.

Reply:
Yeah that makes sense.